File: 039_combination_sum/Solution.go
package main

import (
    "fmt"
    "sort"
)

func combinationSum(candidates []int, target int) [][]int {
    arrLen := len(candidates)
    var results [][]int
    if arrLen == 0 {
        return results
    }

    var path []int
    sort.Ints(candidates)
    dfs(candidates, 0, arrLen, target, results, path)
    return results
}

func dfs(candidates []int, begin, arrLen, target int, results [][]int, path []int) {
    if target == 0 {
        results = append(results, path)
        return
    }

    for i := begin; i < arrLen; i++ {
        if target - candidates[i] < 0 {
            break
        }

        path = append(path, candidates[i])
        dfs(candidates, i, arrLen, target - candidates[i], results, path)
        path = path[:len(path) - 1]
    }
}

func main() {
    nums := []int{2, 3, 6, 7}
    target := 7
    results := combinationSum(nums, target)
    for _, result := range results {
        for _, ele := range result {
            fmt.Printf("%d ", ele)
        }
        fmt.Println();
    }
}

File: 039_combination_sum/Solution.py
class Solution:
    def __init__(self):
        self.path = []
        self.results = []

    def combinationSum(self, candidates: list, target: int) -> list:
        arr_len = len(candidates)
        if arr_len == 0: return self.results

        candidates.sort()
        self.dfs(candidates, 0, arr_len, target)

        return self.results

    def dfs(self, candidates: list, begin: int, arr_len: int, target: int) -> None:
        if target == 0:
            self.results.append(self.path[:]) # 注意此处 path 需要深拷贝
            return

        for i in range(begin, arr_len):
            if target - candidates[i] < 0: break
            self.path.append(candidates[i])
            self.dfs(candidates, i, arr_len, target - candidates[i])
            self.path.pop()


if __name__ == '__main__':
    solution = Solution()
    results = solution.combinationSum([2, 3, 6, 7], 7)
    for result in results:
        print(result)

